docs/languages/csharp-iterators.md

```diff
@@ -0,0 +1,247 @@
+    
+    index = 100;
+    while (index++ < 110)
+        yield return index;
+}
+```
+
+There is one important restriction on iterator methods: you can't have both a
+`return` statement and a `yield return` statement in the same method. The following
+will not compile:
+
+```cs
+public IEnumerable<int> GetSingleDigitNumbers()
+{
+    int index = 0;
+    while (index++ < 10)
+        yield return index;
+        
+    yield return 50;
+   
+    // generates a compile time error: 
+    var items = new int[] {100, 101, 102, 103, 104, 105, 106, 107, 108, 109 };
+    return items;  
+}
+```
+
+This restriction normally isn't a problem. You have a choice of either using
+`yield return` throughout the method, or separating the original method into
+multiple methods, some using `return`, and some using `yield return`.
+
+You can modify the last method slightly to use `yield return` everywhere:
+
+```cs
+public IEnumerable<int> GetSingleDigitNumbers()
+{
+    int index = 0;
+    while (index++ < 10)
+        yield return index;
+        
+    yield return 50;
+   
+    var items = new int[] {100, 101, 102, 103, 104, 105, 106, 107, 108, 109 };
+    foreach (var item in items)
+        yield return item;
+}
+```
+ 
+Sometimes, the right answer is to split an iterator method into two different
+methods. One that uses `return`, and a second that uses `yield return`. Consider
+a situation where you might want to return an empty collection, or the first 5
+odd numbers, based on a boolean argument. You could write that as these two
+methods:
+
+```cs
+public IEnumerable<int> GetSingleDigitOddNumbers(bool getCollection)
+{
+    if (getCollection == false)
+        return new int[];
+    else
+        return IteratorMethod();
+}
+
+private IEnumerable<int> IteratorMethod()
+{
+    int index = 0;
+    while (index++ < 10)
+        if (index % 2 == 1)
+            yield return index;
+}
+```
+ 
+Look at the methods above. The first uses the standard `return` statement to return
+either an empy collection, or the iterator created by the second method. The second
+method uses the `yield return` statement to create the requested sequence.
+
+# Deeper Dive into `foreach`
+
+The `foreach` statement expands into a standard idiom that uses the
+`IEnumable<T>` and `IEnumerator<T>` interfaces to iterate across all
+elements of a colleciton. It also  minimizes errors developers make
+by not properly managing resources. 
+
+The compiler translates the `foreach` loop shown in the first
+example into something similar to this construct:
+
+```cs
+IEnumerator<int> enumerator = collection.GetEnumerator();
+while (enumerator.MoveNext())
+{
+    var item = enumerator.Current;
+    Console.WriteLine(item.ToString());
+}
+```
+
+The construct above represents the code generated by the C# compiler as of
+version 5 and above. Prior to version 5, the `item` variable had a different scope:
+
+```cs
+// C# versions 1 through 4:
+IEnumerator<int> enumerator = collection.GetEnumerator();
```

Comment:
I think we should, because the syntax is the same, but the compiler behavior is different. This was a breaking change, and there is content online that describes the old behavior. I thought that the only way to avoid any confusion would be to describe what it does not, and what the previous behavior was.  See: http://blogs.msdn.com/b/ericlippert/archive/2009/11/12/closing-over-the-loop-variable-considered-harmful.aspx and http://stackoverflow.com/questions/12112881/has-foreachs-use-of-variables-been-changed-in-c-sharp-5